### PR TITLE
documents: replace identifiedBy subjects with $ref

### PR DESCRIPTION
**Pull request by rerowep**
* Implements ABC class for identifiedBy replacement.
* Rewrites contribution identifiedBy replacement.
* Implements identifiedBy subjects replacement for the types `bf:Person` and `bf:Organisation`

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- Which task/US does it implement?
- Which issue does it fix?

## Dependencies

My PR depends on the following `rero-ils-ui`'s PR(s):

* rero/rero-ils-ui#<xx>

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: rero_ils/modules/documents/utils_mef.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+class ReplaceMefIdentifiedByContribution(ReplaceMefIdentifiedBy):
+    """Replace MEF identified by contribution class."""
+
+    def __init__(self, run=False, verbose=False):
+        """Constructor."""
+        self.mef_url = current_app.config.get('RERO_ILS_MEF_AGENTS_URL')
+        self.query = DocumentsSearch() \
+            .filter('exists', field='contribution.agent.identifiedBy')
+        self.cont_types = ['idref', 'gnd']
+        super().__init__(run=run, verbose=verbose)
```
Comment (zannkukai) [inline rero_ils/modules/documents/utils_mef.py:124]: according to previous comment : 
```python
def __init__(self, run=False, verbose=False):
    mef_url = current_app.config.get('RERO_ILS_MEF_AGENTS_URL')
    super().__init__(mef_url, self._query_filter(), run=run, verbose=verbose)
    self.cont_types = ['idref', 'gnd']

def _query_filter():  # could be override by subclasses
     return DocumentsSearch() \
            .filter('exists', field='contribution.agent.identifiedBy')
```
It's in the best practice to call super as soon as possible. This is why I will set `cont_types` after `super` call.

File: rero_ils/modules/documents/utils_mef.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+            click.echo(f'Found: {self.query.count()}')
+        for hit in self.query.source('pid').scan():
+            doc = Document.get_record_by_id(hit.meta.id)
+            new_contributions = []
+            changed = False
+            for contribution in doc.get('contribution', []):
+                new_contributions.append(contribution)
+                ref_type = contribution['agent'].get(
+                    'identifiedBy', {}).get('type', '').lower()
+                if ref_type in self.cont_types + ['rero']:
```
Comment (zannkukai) [inline rero_ils/modules/documents/utils_mef.py:163]: Why don't include `rero` when set up the `self.cont_types` array (into __init__ method) ?

File: rero_ils/modules/documents/utils_mef.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    def __init__(self, run=False, verbose=False):
+        """Constructor."""
+        self.query = DocumentsSearch() \
+            .filter('exists', field='subjects.identifiedBy') \
+            .filter('bool', should=[
+                Q('term', subjects__type='bf:Person'),
+                Q('term', subjects__type='bf:Organisation')
+            ])
+        super().__init__(run=run, verbose=verbose)
```
Comment (zannkukai) [inline rero_ils/modules/documents/utils_mef.py:199]: according to previous review, just can just override the `def _query_filter()` method
```python
def _query_filter():  # could be override by subclasses
     return DocumentsSearch() \
            .filter('exists', field='subjects.identifiedBy') \
            .filter('bool', should=[
                Q('term', subjects__type='bf:Person'),
                Q('term', subjects__type='bf:Organisation')
            ])
```

File: rero_ils/modules/documents/utils_mef.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            click.echo(f'Found: {self.query.count()}')
+        for hit in self.query.source('pid').scan():
+            doc = Document.get_record_by_id(hit.meta.id)
+            new_subjects = []
+            changed = False
+            for subject in doc.get('subjects', []):
+                new_subjects.append(subject)
+                ref_type = subject.get(
+                    'identifiedBy', {}).get('type', '').lower()
+                if ref_type in self.cont_types + ['rero']:
```
Comment (zannkukai) [inline rero_ils/modules/documents/utils_mef.py:213]: same here

**Review by BadrAly — APPROVED** [review]

**Review by vgranata — APPROVED** [review]

File: rero_ils/modules/documents/utils_mef.py (hunk 2 above)
Comment (rerowep) [reply rero_ils/modules/documents/utils_mef.py:163]: We will find RERO identifiedBy but we will replace it only with GND or IDREF $refs.

**Review by zannkukai — APPROVED** [review]

**Review by lauren-d — APPROVED** [review]

**Pull request merged**